Provide a way to shutdown the udev loop gracefully

The project documentation shows a way to use `uevent` library as:

```
dec := uevent.NewDecoder(r)

for {
    evt, err := dec.Decode()
    if err != nil {
	log.Fatal(err)
    }
    fmt.Println(evt)
}
```

I have an app with a goroutine that does this loop. Now I need to stop my application and shutdown the loop. Unfortunately there is no way to do it gracefully. I can do `uevent.Close()` in another goroutine but it is racy and causes a panic in `bufio.Reader`  (see https://github.com/anatol/booster/issues/22):

```
panic: bufio: reader returned negative count from Read

goroutine 6 [running]:
bufio.(*Reader).fill(0xc000070f68)
	/usr/lib/go/src/bufio/bufio.go:103 +0x1dd
bufio.(*Reader).ReadSlice(0xc000070f68, 0x0, 0xc, 0x7f6718544c28, 0xc0003c42f0, 0x0, 0xab0fc0)
	/usr/lib/go/src/bufio/bufio.go:360 +0x3d
bufio.(*Reader).collectFragments(0xc000070f68, 0xc000070d00, 0x40d610, 0xc00006b440, 0x30, 0x30, 0x823120, 0xc00005c960, 0x60, 0x7f671854e308, ...)
	/usr/lib/go/src/bufio/bufio.go:435 +0x7a
bufio.(*Reader).ReadString(0xc000070f68, 0xc00005c900, 0x7, 0x0, 0xc00011a9d8, 0x0)
	/usr/lib/go/src/bufio/bufio.go:483 +0x4c
github.com/s-urbaniak/uevent.(*Decoder).next(...)
	/home/anatol/go/pkg/mod/github.com/s-urbaniak/uevent@v1.0.0/decoder.go:81
github.com/s-urbaniak/uevent.(*Decoder).Decode(0xc000070ed0, 0xc00006b3e0, 0x83cf0f, 0x7)
	/home/anatol/go/pkg/mod/github.com/s-urbaniak/uevent@v1.0.0/decoder.go:42 +0x75
main.udevListener()
	/mnt/cold/sources/golang/booster/init/main.go:454 +0x185
created by main.boost
	/mnt/cold/sources/golang/booster/init/main.go:782 +0x4bd
```

Please provide a way to shutdown the uevent loop gracefully.

--- a/reader.go
+++ b/reader.go
@@ -14,18 +14,36 @@ const NETLINK_KOBJECT_UEVENT = 15
 
 // Reader implements reading uevents from an AF_NETLINK socket.
 type Reader struct {
-	fd int // the file descriptor of the socket.
+	fd     int // the file descriptor of the socket.
+	closed bool
 }
 
 var _ io.ReadCloser = (*Reader)(nil)
 
 // Read reads from the underlying netlink socket.
-func (r Reader) Read(p []byte) (n int, err error) {
-	return syscall.Read(r.fd, p)
+// Trying to read from a closed reader return io.EOF.
+func (r *Reader) Read(p []byte) (n int, err error) {
+	n, err = syscall.Read(r.fd, p)
+	// If the underlying socket has been closed with Reader.Close()
+	// syscall.Read() returns a -1 and an EBADF error.
+	// This Read() function is called by bufio.Reader.ReadString() that
+	// panics if a negative number of read bytes is returned.
+	// Since the EBADF errors could either mean that the file
+	// descriptor is not valid or not open for reading we keep track
+	// if it's actually closed or not and return an io.EOF.
+	if r.closed {
+		return 0, io.EOF
+	}
+	return
 }
 
 // Close closes the underlying netlink socket.
-func (r Reader) Close() error {
+func (r *Reader) Close() error {
+	if r.closed {
+		// Already closed, nothing to do
+		return nil
+	}
+	r.closed = true
 	return syscall.Close(r.fd)
 }
 
@@ -50,5 +68,5 @@ func NewReader() (io.ReadCloser, error) {
 	}
 
 	err = syscall.Bind(fd, &nl)
-	return &Reader{fd}, err
+	return &Reader{fd: fd}, err
 }

